Upgrade to Symfony 4

alanhartless opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | 4.0
| Bug fix?                               | no
| New feature?                           | no
| Deprecations?                          | yes
| BC breaks?                             | yes
| Automated tests included?              | some
| Related user documentation PR URL      | n/a
| Related developer documentation PR URL | todo

#### Description:

This is a PR that upgrades Mautic to Symfony 4 and minimal changes to make it "work." Please note the UPGRADE-4.0.md doc. 

There are some notes in https://docs.google.com/document/d/1VNQ5cuXJdzsgxi-HTfZ_ihSVrNBSWZwgClG6pabY7nc/edit#heading=h.2v7yth9cwgwg. A formal project/plan will be coming soon. 

Changes so far:
* Upgrade Symfony 4 (required some "hacks" to make services public and auto-register Commands; preferable these get refactored but will track in the planning board)
* Upgraded to latest El Finder (required to remove abandoned packages)
* Requires https://github.com/mautic/BazingaOAuthServerBundle/tree/release-0.0.4 for OAuth1a compatibility
* Refactored ModeratedCommand (Symfony removed LockHandler)
* Refactored some things in PipeDrive integration
* Upgraded GuzzleHttp to v7
* Bumped minimal PHP version to 7.4 - we're planning to add support for PHP 8.0 later on

## How to test

It's easiest if you do a **clean checkout** of this PR, e.g. in a new folder:

1. `gh repo clone mautic/mautic && cd mautic && gh pr checkout 9409`
2. If you have DDEV, simply run `ddev start` to get up and running, it will do the installation for you as well!
3. Mautic should install without errors. You should be able to click around, but don't expect all features to work flawlessly just yet, we haven't tested them.

alanhartless committed:
Throw an exception if the run directory cannot be created rather than failing silently

alanhartless committed:
Replaced DefinitionDecorator with ChildDefinition

alanhartless committed:
Updated generateUrl signature to match the new signature in Symfony

alanhartless committed:
Set services to be public by default as expected by M3.

alanhartless committed:
Replaced ContextErrorException with \ErrorException

alanhartless committed:
Set CommonEntity as a super class

alanhartless committed:
Replaced Twig_Extension with AbstractExtension and Twig_SimpleFunction with TwigFunction

alanhartless committed:
Method must be an empty array for Symfony to now allow all methods (empty string results in no methods allowed)

alanhartless committed:
Replaced TokenStorage with TokenStorageInterface

alanhartless committed:
Remove now unsupported RSS reader that has been disabled for a long time now

alanhartless committed:
Update Symfony to 4.4, remove unsupported bundles, and update willdurand/oauth-server-bundle to release-0.0.4 dev branch on mautic/BazingaOAuthServerBundle that has just enough Symfony 4 changes to at least boot Mautic

alanhartless committed:
Update composer.lock

alanhartless committed:
Bump PHP min to 7.3

alanhartless committed:
Remove codeception

alanhartless committed:
Replaced abandoned piwik/device-detector with matomo/device-detector

alanhartless committed:
Must call isSubmitted before isValid

alanhartless committed:
Upgrade ElFinder bundle to remove abandoned robloach/component-installer

alanhartless committed:
Set some commonly used services as public (hack)

alanhartless committed:
Refactored ModeratedCommand to use the Lock component as Symfony removed LockHandler

alanhartless committed:
Replaced deprecated namespaces for ArraySourceIterator and IteratorSourceIterator

alanhartless committed:
Remove defunct translation debug command

alanhartless committed:
Autowire commands

alanhartless committed:
Symfony's container is case sensitive now

alanhartless committed:
Symfony requires controllers as services to be tagged with controller.service_arguments and uses the classname as the service name. Made aliases public for BC with M3.

alanhartless committed:
Start of M4 upgrade notes

alanhartless committed:
Removed deprecations removed in Symfony 4

alanhartless committed:
Updated Symfony 4 compatibility of Bazinga oauth1a

alanhartless committed:
Note on new M4 hacks

alanhartless committed:
Replaced DeviceParserAbstract with AbstractDeviceParser

dennisameling committed:
Add support for Composer v2 🚀

dennisameling committed:
Remove PHP 7.2 from CI

dennisameling commented:
Just added https://github.com/mautic/mautic/pull/9409/commits/902b53bd5a1258c6cd992740dc5a3a22a968ae78 which adds support for Composer v2 🚀 cc @RCheesley

dennisameling commented:
Given the expected release date of M4, what do we think about bumping the minimum PHP version to 7.4? Would also allow us to start using things like typed properties: https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.typed-properties 🚀 @mautic/core-team

escopecz commented:
@dennisameling yes, we've discussed it internally. PHP 7.4 should be the minimum version for Mautic 4 as the older versions won't get security fixes anymore during Mautic 4 lifetime.

dennisameling committed:
Bump minimum PHP version to 7.4

dennisameling committed:
Update composer.lock

dennisameling committed:
Don't enable PHP 8.0 on CI just yet 🤓

dennisameling committed:
Fix PHP Fatal error:  Cannot redeclare static Symfony\Bundle\FrameworkBundle\Test\WebTestCase::$container

dennisameling committed:
Fix CS

dennisameling committed:
Merge remote-tracking branch 'upstream/features' into m4-quick-and-dirty

dennisameling committed:
Fix fatal PHP errors

dennisameling commented:
Just merged the latest changes from the `features` branch into this one (which includes the PHPUnit 8 upgrade for PHP 7.4+ support). Fixed the fatal PHP errors (https://github.com/mautic/mautic/pull/9409/commits/d655ca12eb6660ee808bed17a24df1590d82fe55), but there's still a long way to go for this PR. We'll get there 🚀

dennisameling committed:
Use Symfony's new static $container variable

dennisameling committed:
Replace $this->container with self::$container in several places

dennisameling committed:
Replace $this->container with self::$container in more places

dennisameling commented:
Woohoo! Can log into the dashboard now. Version is v4.0.0-alpha1 🚀 

![image](https://user-images.githubusercontent.com/17739158/105360093-810c0e00-5bf8-11eb-995b-9fab40013a61.png)

Surprisingly little PHPUnit errors (33x) and failures (6); it should be possible to fix those quite quickly.

There are some PHP Notices/warnings in there which will be fixed by https://github.com/mautic/mautic/pull/9602 - let's focus on that PR first, merge it into this one, and then continue work here 👍🏼

dennisameling committed:
Merge remote-tracking branch 'upstream/features' into m4-quick-and-dirty

dennisameling committed:
Re-enable PsyshBundle after PHPUnit 9 upgrade

dennisameling commented:
PHPUnit 9 upgrade complete including several fixes for PHP Notices/warnings (https://github.com/mautic/mautic/pull/9602). Just merged it into this PR, so we can continue working on this one again 👍🏼 

**This PR is open for anyone who has some time to work on it. First step would be to get the PHPUnit tests to pass.**

dennisameling committed:
Merge remote-tracking branch 'upstream/features' into m4-quick-and-dirty

dennisameling commented:
Just merged the latest changes from the `features` branch into this one 👍🏼 

_Note to self:_ installer is broken due to `secret_key` being empty in a new installation. Need to fix that soon, but want to wait for https://github.com/mautic/plugin-grapesjs-builder/issues/59 to be fixed first, as that one also breaks new installations

dennisameling committed:
Merge remote-tracking branch 'upstream/features' into m4-quick-and-dirty

dennisameling commented:
Just merged the latest changes from the `features` branch into this one, which includes the fix for https://github.com/mautic/plugin-grapesjs-builder/issues/59. 

Have blocked 2 hours on March 7 to continue looking into the failing tests

dennisameling committed:
Fix several tests

dennisameling commented:
Temporarily updated base branch to `features` so it's easier to compare the changes we've done so far (since the `4.0` branch is quite far behind at this point and will need to be updated at some point)

dennisameling committed:
Fix another test failure

dennisameling committed:
Add Symfony's HttpClient + CSS selector deps

dennisameling commented on app/bundles/CoreBundle/Config/config.php:
```diff
@@ -1187,7 +1190,7 @@
         'db_server_version'               => '5.7',
         'locale'                          => 'en_US',
-        'secret_key'                      => '',
+        'secret_key'                      => 'temp',
```
Note to self: needed this to get the installer to work (since no `secret_key` is set at that point yet). Might have to find a different solution here; discuss in #mautic-4 Slack channel later

dennisameling replied:
Have added this to the Mautic 4 JIRA board for looking into it later: https://mautic.atlassian.net/browse/TPROD-169

dennisameling committed:
Refactor ConsoleExceptionListener to ConsoleErrorListener

Symfony 4.0 removed console.exception in favor of console.error. Details: https://github.com/symfony/symfony/blob/4.4/UPGRADE-4.0.md

dennisameling committed:
Upgrade Guzzle to v7

Guzzle 7 implements PSR-18, which has several benefits. We can e.g. remove the dependency on ricardofiorani/guzzle-psr18-adapter thanks to this!

dennisameling committed:
Set KERNEL_CLASS env variable in PHPUnit config

From the Symfony 4.0 upgrade docs (https://github.com/symfony/symfony/blob/4.4/UPGRADE-4.0.md):

Set the KERNEL_CLASS environment variable to the fully-qualified class name of your Kernel or override the KernelTestCase::createKernel() or KernelTestCase::getKernelClass() method instead.

dennisameling committed:
Fix undefined variable: $containerAttributesText

dennisameling committed:
Refactor mautic.webhook.http.client to use mautic.http.client instead of mautic.guzzle.client

dennisameling committed:
Update Composer lockfile

dennisameling committed:
Fix Rector by downgrading PHPSTAN

dennisameling committed:
FIx PHPSTAN errors

dennisameling committed:
Don't fail fast in CI

dennisameling committed:
Fix CS

dennisameling committed:
Fix SparkPost tests

dennisameling committed:
Fix two tests

dennisameling committed:
Fix EmailControllerFunctionalTest

dennisameling committed:
Upload test logs as artifacts

dennisameling committed:
Fix two tests

dennisameling committed:
Fix Pipedrive tests

dennisameling committed:
Fix webhookbundle ClientTest

dennisameling commented:
Note to self: some PHPUnit tests are failing which are related to CLI commands:

```
2021-03-21T11:42:52.8151452Z 28) MauticPlugin\MauticCrmBundle\Tests\Pipedrive\Command\FetchPipedriveDataCommandTest::testCommandWithdisableIntegration
2021-03-21T11:42:52.8152824Z Failed asserting that 0 matches expected 1.
2021-03-21T11:42:52.8153203Z 
2021-03-21T11:42:52.8154185Z /home/runner/work/mautic/mautic/plugins/MauticCrmBundle/Tests/Pipedrive/Command/FetchPipedriveDataCommandTest.php:48
2021-03-21T11:42:52.8155092Z 
2021-03-21T11:42:52.8155998Z 29) MauticPlugin\MauticCrmBundle\Tests\Pipedrive\Command\FetchPipedriveDataCommandTest::testCommand
2021-03-21T11:42:52.8157344Z Failed asserting that 2 matches expected 3.
2021-03-21T11:42:52.8157904Z 
2021-03-21T11:42:52.8158986Z /home/runner/work/mautic/mautic/plugins/MauticCrmBundle/Tests/Pipedrive/Command/FetchPipedriveDataCommandTest.php:86
```

When running these commands directly on the CLI (setting `--env=test `), the actual errors show up:

```
denni@mautic4-web:/var/www/html$ SYMFONY_ENV=test MAUTIC_ENV=test bin/console --env=test mautic:integration:pipedrive:fetch
12:00:08 ERROR     [console] Error thrown while running command "--env=test mautic:integration:pipedrive:fetch". Message: "Call to a member function getIntegrationSettings() on bool" ["exception" => Error { …},"command" => "--env=test mautic:integration:pipedrive:fetch","message" => "Call to a member function getIntegrationSettings() on bool"]
12:00:08 WARNING   [mautic] Command `mautic:integration:pipedrive:fetch` exited with status code 1

In FetchPipedriveDataCommand.php line 48:
                                                              
  Call to a member function getIntegrationSettings() on bool  
                                                              

mautic:integration:pipedrive:fetch [--restart] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command>
```

Ideally, we'd need to find a way to let those errors show up in the PHPUnit output and/or our logs, for easier debugging in the future.

dennisameling committed:
Refactor Pipedrive commands to use proper DI

dennisameling committed:
Explicitly set some services to public in test config

dennisameling committed:
Temp install Mautic prior to running tests

Some tests are failing if Mautic isn't installed first. This will need more investigation later, but this should fix things as a workaround.

dennisameling committed:
Set explicit property types on $this->client in AbstractMauticTestCase

dennisameling committed:
Check if $this->client is set

dennisameling committed:
Fix several failing tests

dennisameling committed:
Run Pipedrive tests separately due to bug

dennisameling commented:
Only ~two more failing tests~! I have to stop here due to time constraints and insufficient knowledge of the two failing modules. If someone could take that on, we'll be ready to merge this! 🚀 

UPDATE: only one more failing test after https://github.com/mautic/mautic/pull/9409/commits/018ede8861f67e437e7c63917116c8aa3762863a 🎉

dennisameling committed:
Add example to upgrade guide about HTTP client

dennisameling committed:
Revert mautic.webhook.http.client to use mautic.guzzle.client again

dennisameling committed:
Merge remote-tracking branch 'upstream/features' into m4-quick-and-dirty

dennisameling committed:
Fix failing tests after merge of latest commits in features branch

dongilbert commented:
Taking this on to address the test failures

dongilbert committed:
Address test failure

dongilbert committed:
Remove special schema handling in test

dongilbert committed:
Remove usage of custom table prefix for RemoveDuplicateIndexDataTest

dennisameling committed:
Fix RemoveDuplicateIndexDataTest

dennisameling committed:
Ugly fix for RemoveDuplicateIndexDataTest

dennisameling committed:
Let's not drop those tables to avoid conflicts with other tests

dennisameling commented on app/phpunit.xml.dist:
```diff
@@ -27,6 +27,7 @@
     </testsuite>
     <testsuite name="Plugin tests">
       <directory>./../plugins/*Bundle/Tests</directory>
+      <exclude>./../plugins/MauticCrmBundle/Tests/Pipedrive</exclude>
```
We run these tests separately in CI now due to a bug. They do pass if they are ran separately. Since we're planning to decouple these plugins from core, I think it's OK for now to temporary have this situation.

dennisameling committed:
Add back post-checkout hooks

dennisameling reviewed (approved):
All failing tests are passing again now. This PR is crucial for further development of Mautic 4. Where applicable, we created issues in the Mautic 4 JIRA board: https://mautic.atlassian.net/secure/RapidBoard.jspa?projectKey=TPROD&rapidView=26&atlOrigin=eyJpIjoiNzRlMTMxMzQxMGMzNGI0MWIyY2EwNTFlNmZhMzA4M2YiLCJwIjoiaiJ9

The [slightly lower Codecov coverage](https://github.com/mautic/mautic/pull/9409#issuecomment-764768523) should be OK for now due to the large amount of changes we did in this PR!

dennisameling commented:
@RCheesley Could we remove the required `PHPUnit 7.3 mariadb` and `PHPUnit 7.3 mysql` checks? PHP 7.4 is the minimum required version now, so we don't test against older versions anymore. Thanks in advance! 😊

RCheesley commented:
Removed on the Features branch - should I also remove on the 3.3 or leave it for that branch?

dennisameling commented:
It's fine to leave it on for that branch, thanks!
